modules/control/submodules/preprocessor_submodule.cc

```diff
@@ -0,0 +1,333 @@
+using apollo::common::Status;
+using apollo::common::VehicleStateProvider;
+using apollo::common::time::Clock;
+using apollo::localization::LocalizationEstimate;
+using apollo::planning::ADCTrajectory;
+
+PreprocessorSubmodule::PreprocessorSubmodule()
+    : monitor_logger_buffer_(common::monitor::MonitorMessageItem::CONTROL) {}
+
+PreprocessorSubmodule::~PreprocessorSubmodule() {}
+
+std::string PreprocessorSubmodule::Name() const {
+  return FLAGS_preprocessor_submodule_name;
+}
+
+bool PreprocessorSubmodule::Init() {
+  if (!cyber::common::GetProtoFromFile(FLAGS_control_common_conf_file,
+                                       &control_common_conf_)) {
+    AERROR << "Unable to load control common conf file: " +
+                  FLAGS_control_common_conf_file;
+    return false;
+  }
+
+  /*  initialize readers and writers */
+  // TODO(SHU): add writer of preprocessor
+  cyber::ReaderConfig chassis_reader_config;
+  chassis_reader_config.channel_name = FLAGS_chassis_topic;
+  chassis_reader_config.pending_queue_size = FLAGS_chassis_pending_queue_size;
+  chassis_reader_ =
+      node_->CreateReader<Chassis>(chassis_reader_config, nullptr);
+  CHECK(chassis_reader_ != nullptr);
+
+  cyber::ReaderConfig planning_reader_config;
+  planning_reader_config.channel_name = FLAGS_planning_trajectory_topic;
+  planning_reader_config.pending_queue_size = FLAGS_planning_pending_queue_size;
+  trajectory_reader_ =
+      node_->CreateReader<ADCTrajectory>(planning_reader_config, nullptr);
+  CHECK(trajectory_reader_ != nullptr);
+
+  cyber::ReaderConfig localization_reader_config;
+  localization_reader_config.channel_name = FLAGS_localization_topic;
+  localization_reader_config.pending_queue_size =
+      FLAGS_localization_pending_queue_size;
+  localization_reader_ = node_->CreateReader<LocalizationEstimate>(
+      localization_reader_config, nullptr);
+  CHECK(localization_reader_ != nullptr);
+
+  cyber::ReaderConfig pad_msg_reader_config;
+  pad_msg_reader_config.channel_name = FLAGS_pad_topic;
+  pad_msg_reader_config.pending_queue_size = FLAGS_pad_msg_pending_queue_size;
+  pad_msg_reader_ =
+      node_->CreateReader<PadMessage>(pad_msg_reader_config, nullptr);
+  CHECK(pad_msg_reader_ != nullptr);
+
+  // set initial vehicle state by cmd
+  // need to sleep, because advertised channel is not ready immediately
+  // simple test shows a short delay of 80 ms or so
+  AINFO << "Control resetting vehicle state, sleeping for 1000 ms ...";
+  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+
+  // should init_vehicle first, let car enter work status, then use status msg
+  // trigger control
+  AINFO << "Control default driving action is: "
+        << DrivingAction_Name(control_common_conf_.action());
+  pad_msg_.set_action(control_common_conf_.action());
+
+  return true;
+}
+
+bool PreprocessorSubmodule::Proc() {
+  chassis_reader_->Observe();
+  const auto &chassis_msg = chassis_reader_->GetLatestObserved();
```

Comment:
It's a shared pointer. Just get a copy instead of reference. So that you share the ownership and use it safely.